src/mito2/src/read/compat.rs

```diff
@@ -156,7 +157,21 @@ impl CompatFields {
             .index_or_defaults
             .iter()
             .map(|index_or_default| match index_or_default {
-                IndexOrDefault::Index(index) => batch.fields()[*index].clone(),
+                IndexOrDefault::Index { pos, cast_type } => {
+                    let old_column = &batch.fields()[*pos];
+
+                    let data = if let Some(ty) = cast_type {
+                        // Safety: We ensure type can be converted and the new batch should be valid.
+                        // Tips: `safe` must be true in `CastOptions`, which will replace the specific value with null when it cannot be converted.
+                        old_column.data.cast(ty).unwrap()
```

Comment:
We need test cases to ensure updating arrow won't break this. We could test some behavior in `VectorOp::cast()`'s test cases.